src/transformers/modeling_utils.py

```diff
@@ -846,20 +843,25 @@ def generate(
             attention_mask = attention_mask.contiguous().view(
                 effective_batch_size * num_beams, input_ids_len
             )  # shape: (batch_size * num_return_sequences * num_beams, cur_len)
+
         if self.config.is_encoder_decoder:
             assert bos_token_id is not None, "Encoder Decoder Models need to have a bos_token_id"
-            # encoder decoder need to start with empty input_ids and copy the input_ids to encoder_inputs
-            encoder_inputs = input_ids
+            assert hasattr(self, "encode"), "{} should have a 'encode' function defined".format(self)
+            assert callable(self.encode), "{} should be a method".format(self.encode)
+
+            # get encoder outputs once and store them
+            encoder_outputs = self.encode(input_ids, attention_mask=attention_mask)
```

Comment:
Can we also have a `get_encoder()` method here (as mentioned above for TF)?

We could actually remove this `encode()` methods which (unless I'm missing something) as just the forward pass of the encoder. I don't see the need for another abstraction here.